[sparse] Add cuSPARSELt as a backend

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134022
* __->__ #128534

Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[sparse] Add cuSPARSELt as a backend

Summary:

This PR adds in cuSPARSELt as a backend to PyTorch

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
cc @albanD and @malfet Sorry, I wasn't quite sure who to request a review from, since Christian is out, feel free to reassign if there's a better person.

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@malfet I added in a test for the cusparselt version [here](https://github.com/pytorch/pytorch/pull/128534/files#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR1111).

We're already bundling cuSPARSELt with the binary [here](https://github.com/pytorch/builder/pull/1672/files), so there shouldn't be any additional 

I think cuSPARSELt is still in the process of adding to pip / conda, we can remove the above code once they finalize.

jcaip commented:
@eqy Can you take a look at this please? Since you're the codeowner for some of these files :)

albanD reviewed (commented):
It's a lot of boiler plate copy pasted code with little benefit for now. But I guess you're going to add more to it?
Could we share some code with the cudnn binding to reduce how much boilerplate we have?

Also curious if you could add this to the cuda backend module like we do for cufft? That would also reduce the amount of boiler plate code

jcaip commented:
Sounds good, let me try to add this to the cuda backend instead.

jcaip commented:
@pytorchbot rebase

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
cc @albanD Sorry, forgot about this 😬 . Revisiting this now as I'm updating cuSPARSELt to the latest version and this would be nice to have. I think I'd like to leave it as a standalone backend for now though. Since we're updating the cuSPARSELt version anyways I plan to move some configs out of the subclass into the backend. 

Also cc @cpuhrsch now that you're back :)

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] Add cuSPARSELt as a backend"


Summary:

This PR adds in cuSPARSELt as a backend to PyTorch.

It is now possible to see if cuSPARSELt is available and the version if
it is with
```
torch.backends.cusparselt.is_available()
torch.backends.cusparselt.version()
```

Test Plan:
```
python test/test_sparse_semi_structured.py -k test_cusparselt_backend
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

syed-ahmed commented on test/test_sparse_semi_structured.py:
```diff
@@ -1113,6 +1108,22 @@ def test_cslt_sparse_mm_search(self, device, dtype):
+        elif version == (12, 1):
+            assert torch.backends.cusparselt.version() == 502
+        elif version > (12, 1):
+            assert torch.backends.cusparselt.version() == 502
```
Do we need to update this number (502) every time we update cuda or cusparselt?

jcaip replied:
Not when CUDA is updated but yes when cuSPARSELt is. I have a follow-up [PR](https://github.com/pytorch/pytorch/pull/134022) to bump it to v0.6.2

syed-ahmed replied:
Since we have to update the number every time we bump cusparselt, because of that I think the test adds some maintenance burden. I would suggest simplifying the test. For instance, if you are just testing the `version()` API integration, may be just testing that it returns a version >= the minimum cusparselt version is sufficient.

syed-ahmed reviewed (commented):
One high level question I have is do you need `torch.backends.cusparselt.is_available()` and `torch.backends.cusparselt.version`? The only place I see a usage is the test, so I'm curious where in pytorch execution would you need to specialize based on cusparselt availability and version?

jcaip commented:
> One high level question I have is do you need `torch.backends.cusparselt.is_available()` and `torch.backends.cusparselt.version`? The only place I see a usage is the test, so I'm curious where in pytorch execution would you need to specialize based on cusparselt availability and version?

@syed-ahmed Different cusparselt versions offer different performance characteristics / feature support, so it's useful to have this function if we get bug reports so we can know what's expected or not. Right now the only way to get that [info](https://github.com/fairinternal/xformers/blob/196ef192be303323ef450907df228659f6235ebc/xformers/ops/sp24.py#L56) is to manually dlopen the library which is kind of hacky.

syed-ahmed commented:
> > One high level question I have is do you need `torch.backends.cusparselt.is_available()` and `torch.backends.cusparselt.version`? The only place I see a usage is the test, so I'm curious where in pytorch execution would you need to specialize based on cusparselt availability and version?
> 
> @syed-ahmed Different cusparselt versions offer different performance characteristics / feature support, so it's useful to have this function if we get bug reports so we can know what's expected or not. Right now the only way to get that [info](https://github.com/fairinternal/xformers/blob/196ef192be303323ef450907df228659f6235ebc/xformers/ops/sp24.py#L56) is to manually dlopen the library which is kind of hacky.

Thanks for the context. That sounds reasonable. In a followup PR, I would suggest add an entry to `collect_env` so that we can improve debuggability as you mentioned.

jcaip commented:
Thanks, will update the collect env script in a future PR

jcaip commented:
@pytorchbotmerge -f "unrelated test failure"

jcaip commented:
@pytorchbot merge -f "unrelated test failure"
